### ISSUE TITLE
error with bonomial.log_pdf

### ISSUE THREAD

**Issue by dori-reichmann** [uber/pyro#675]
A simple attempt to sample, and calculate the pdf:
```
import torch
from torch.autograd import Variable
import pyro
import pyro.distributions as dist

theta = Variable(torch.Tensor([0.2,0.3,0.5]))
n = Variable(torch.Tensor([100]))
clicks = pyro.sample("measurement", dist.multinomial, theta,n)
log_pdf_clicks = dist.multinomial.log_pdf(clicks,theta,n)
```

And here it the error:

> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-2-72ed00f2b7a4> in <module>()
>       7 n = Variable(torch.Tensor([100]))
>       8 clicks = pyro.sample("measurement", dist.multinomial, theta,n)
> ----> 9 log_pdf_clicks = dist.multinomial.log_pdf(clicks,theta,n)
> 
> /opt/miniconda3/lib/python3.6/site-packages/pyro/distributions/random_primitive.py in log_pdf(self, x, *args, **kwargs)
>      40 
>      41     def log_pdf(self, x, *args, **kwargs):
> ---> 42         return self.dist_class(*args, **kwargs).log_pdf(x)
>      43 
>      44     def batch_log_pdf(self, x, *args, **kwargs):
> 
> /opt/miniconda3/lib/python3.6/site-packages/pyro/distributions/distribution.py in log_pdf(self, x, *args, **kwargs)
>     183         :rtype: torch.autograd.Variable
>     184         """
> --> 185         return torch.sum(self.batch_log_pdf(x, *args, **kwargs))
>     186 
>     187     @abstractmethod
> 
> /opt/miniconda3/lib/python3.6/site-packages/pyro/distributions/multinomial.py in batch_log_pdf(self, x)
>      84         """
>      85         batch_log_pdf_shape = self.batch_shape(x) + (1,)
> ---> 86         log_factorial_n = log_gamma(x.sum(-1) + 1)
>      87         log_factorial_xs = log_gamma(x + 1).sum(-1)
>      88         log_powers = (x * torch.log(self.ps)).sum(-1)
> 
> /opt/miniconda3/lib/python3.6/site-packages/pyro/distributions/util.py in log_gamma(xx)
>      22     x = xx - 1.0
>      23     t = x + 5.5
> ---> 24     t = t - (x + 0.5) * torch.log(t)
>      25     ser = Variable(torch.ones(x.size()).type(ttype)) * magic1
>      26     for c in gamma_coeff:
> 
> RuntimeError: log is not implemented for type torch.LongTensor

I am running with
(python) sys.version = '3.6.4 |Anaconda, Inc.| (default, Dec 21 2017, 21:42:08) \n[GCC 7.2.0]'
torch.__version__ = '0.3.0.post4'
pyro.__version__ = '0.1.2'

**Comment by fritzo** [comment]
Thanks for the detailed bur report. We'll be switching to [PyTorch distributions](http://pytorch.org/docs/master/distributions.html) in a few days. When we do, we'll ensure that the bug is fixed. I'll leave this issue open until we migrate.

**Comment by 1Reinier** [comment]
I ran into this too. As a workaround you can cast Tensor `n` with .`float()`. 

@fritzo Note that in the Pytroch distributions Pyro intends to switch to, multinomial does not accept Tensors for `n` (just numbers). It would be nice if we can batch that operation too in Pytorch.

**Comment by fritzo** [comment]
@Reinier, what are your use cases for heterogeneous `n` in `Multinomial`? You're correct that the PyTorch version accepts only a single `n`. However that `n` is not required if the multinomial is used only for `.log_prob()` calls. Therefoer the PyTorch `Multinomial` should support heterogeneous `n` if it is only used in Pyro observe statements e.g. observing word counts in a batch of documents where each document has a different length.

**Comment by 1Reinier** [comment]
Yeah exactly, I'm using it in an observe statement.

**Comment by dori-reichmann** [comment]
great, thanks

**Comment by 1Reinier** [comment]
@fritzo quick question (don't want to reopen this) about broadcasting the PyTorch Multinomial distribution over `total_counts`: if I leave the argument default (meaning == 1), the observe statement will work as expected (even though the counts are different per row?). Thanks!

**Comment by fritzo** [comment]
@1Reinier Yes `torch.Multinomial(probs=probs).log_prob(counts)` should support heterogeneous counts. We simply haven't implemented a heterogeneous sampler. cc @neerajprad this might be useful since you can use heterogeneous `Multinomial` in observe statements and in HMC.

**Comment by neerajprad** [comment]
@fritzo - Thanks for pointing this out. This works for the example but is quite a bit slower than using Binomial, unfortunately.

**Comment by gewirtz** [comment]
Hi, @fritzo I can't get the Multinomial to work with heterogenous counts like you said it could above. Here is a basic example:

import torch
import torch.distributions as tdist

x = torch.tensor([[3.,0.,0.], [0.,2.,0.], [1.,2.,1.]])
x2 = torch.tensor([[1.,0.,0.], [0.,1.,0.], [0.,0.,1.]])
p = torch.tensor([[0.99,0.,0.01], [0.95,0.05,0.], [1./3.,1./3.,1./3.]])

The following works, where the total counts are all 1:
tdist.Multinomial(probs=p).log_prob(x2)

However, this yields an error:
tdist.Multinomial(probs=p).log_prob(x)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-61-e6d65e933790> in <module>
      4 p2 = torch.ones([3]) * (1.0/3.0)
      5 print(tdist.Multinomial(probs=p).log_prob(x2))
----> 6 tdist.Multinomial(probs=p).log_prob(x)

~/.local/lib/python3.7/site-packages/torch/distributions/multinomial.py in log_prob(self, value)
    101     def log_prob(self, value):
    102         if self._validate_args:
--> 103             self._validate_sample(value)
    104         logits, value = broadcast_all(self.logits.clone(memory_format=torch.contiguous_format), value)
    105         log_factorial_n = torch.lgamma(value.sum(-1) + 1)

~/.local/lib/python3.7/site-packages/torch/distributions/distribution.py in _validate_sample(self, value)
    251 
    252         if not self.support.check(value).all():
--> 253             raise ValueError('The value argument must be within the support')
    254 
    255     def _get_checked_instance(self, cls, _instance=None):

ValueError: The value argument must be within the support


I have an N x P data matrix that is multinomial, but each document has a different total word count. I'm at a loss as to how to get this to work in a model observation statement. Any help would be GREATLY GREATLY appreciated!! Thank you!

**Comment by fritzo** [comment]
@gewirtz you can either set `total_count` to an upper bound on the number of counts via `dist.Multinomial(probs=p, total_count=999)`, or disable validation via `dist.Multinomial(probs=p, validate_args=False)`.

**Comment by gewirtz** [comment]
Thank you so much!!
